options/cf_options.cc

```diff
@@ -658,19 +658,12 @@ static std::unordered_map<std::string, OptionTypeInfo>
                    // it's a const pointer of const Comparator*
                    const auto* ptr =
                        static_cast<const Comparator* const*>(addr);
-                   // Since the user-specified comparator will be wrapped by
-                   // InternalKeyComparator, we should persist the
-                   // user-specified one instead of InternalKeyComparator.
                    if (*ptr == nullptr) {
                      *value = kNullptrString;
                    } else if (opts.mutable_options_only) {
                      *value = "";
                    } else {
-                     const Comparator* root_comp = (*ptr)->GetRootComparator();
-                     if (root_comp == nullptr) {
-                       root_comp = (*ptr);
-                     }
-                     *value = root_comp->ToString(opts);
+                     *value = (*ptr)->ToString(opts);
```

Comment:
Trying to think about if this can affect existing users:
* If they use built-in comparators:
  * if no user-defined timestamp, behavior is the same as before.
  * if they user-defined timestamp, BytewiseComparatorWithU64Ts will be set instead of BytewiseComparator, this is probably fine since udt is experimental.
* If they use customized comparators, they would need to manually set comparator already.

What do you think?

Reply:
We have four builtin comparators: `BytewiseComparator`, `ReverseBytewiseComparator`, `BytewiseComparatorWithU64Ts`, ReverseBytewiseComparatorWithU64Ts`. 

For the former two: `GetRootComparator` returns `this`. So the behavior of the current code is the same as with this change.

For the latter two, the current behavior is the root comparator `BytewiseComparator` and `ReverseBytewiseComparator` will be recorded in OPTIONS file, instead of the `*WithU64Ts` version. This behavior I think is an incorrect one. If any existing users are parsing from OPTIONS file and pass it to DB::Open, they will likely run into the same error message as the one in this stress test: T168583304

If user is using a customized comparator, and their implementation is not overriding the `GetRootComparator` API, the behavior with and without this change is the same:
https://github.com/facebook/rocksdb/blob/8e1adab5cecad129131a4eceabe645b9442acb9c/include/rocksdb/comparator.h#L119 If their implementation is overriding `GetRootComparator` API, I still doubt that they really want the root comparator recorded in OPTIONS file instead of the actual comparator. If they are parsing OPTIONS file, I think they will run into error like above stress test did.

There is a VersionEdit recorded in Manifest for each column family's comparator name, that record will have the comparator's name instead of its root comparator's name so there is a discrepancy:
https://github.com/facebook/rocksdb/blob/8e1adab5cecad129131a4eceabe645b9442acb9c/db/db_impl/db_impl.cc#L3533 They should be aligned IMHO.

Reply:
Sg, I don't see any override for GetRootComparator() from a quick search.